software-engineering-policies/Containers/ContainerBestPracticies.md

```diff
@@ -0,0 +1,51 @@
+
+Docker provide a series of best practices for writing and developing Dockerfile’s. These are:
+
+* Best practices for [writing a Dockerfile](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/).
+* Best practices for [developing with Docker](https://docs.docker.com/develop/dev-best-practices/).
+* Best practices for [secure development with Docker](https://docs.docker.com/develop/security-best-practices/).
+
+The following items supersede these best practices:
+
+* Where a container needs to use secrets and passwords, these should be injected into the container via environment variables defined in the Dockerfile and should be encrypted at a minimum.
+* Dockerfile’s should be linted to detect issues and vulnerabilities
+* Aim to use minimal dependencies, such as [distroless container base](https://github.com/GoogleContainerTools/distroless).
+* Avoid defining unnecessary privileges and avoid running as root.
+* Ensure there is a well-established process for updating dependencies.
+* Where both suitable and possible, leverage [multi-stage builds](https://docs.docker.com/build/building/multi-stage).
+* The UK Hydrographic office have a number of [approved images at DockerHub](https://hub.docker.com/u/ukhydrographicoffice) and it is recommended new images are based on these. The Dockerfile files for these images are stored within the [UKHO dockerimages repo on GitHub](https://github.com/UKHO/dockerimages).
+
+## Container Deployment Best Practices
+
+### Must Do
+
+* Where applicable at deployment, secrets must be stored via Azure KeyVault, AWS Secrets Manager, etc.
+* The containers resources, limits and requests for CPU and Memory consumption must be defined for deployment
+* Containers deployments and deployment environment must be configured using infrastructure as code.
+* Deployments must include a means to monitor the health of the container. For example, when using Kubernetes, configure readiness and liveliness probes.
+* Environments above Dev must have HTTPS enforced and HTTP to HTTPS redirects enabled.
+* Logs generated by containers must be consumed.
+* Access to the underlying infrastructure running the container images must be blocked in all environments above Dev.
+* Container images must be scanned by a vulnerability scanning tool.
+* Container deployments must not become a persistance data store or silo.
+
+### Should do
+
+* Microservices and containers should communicate only with services and ports they need to.
+* Access Policies and Network Security Groups should be applied.
+* Container deployments should be isolated using Namespaces.
+* Monitoring solution should be in place for deployed containers. For example, ELK (.Net), EFK (Java) stacks, Grafana, Prometheus etc.
+* Automated deployments should be used on container image version increments within the container registry.
+
+### Could do
+
+* DAPR sidecar could be utilised for third party component management (queues, DB's etc).
```

Comment:
This is interesting one. Do we use DAPR anywhere at UKHO? We don't currently deploy the operator to our data platform clusters but I'd like to if it's an approved tool.

Reply:
It's not listed on the tech radar https://ukhotechradarweblive.z33.web.core.windows.net/

Reply:
I dropped this to a "Could", since I was not sure we utilised it. Is this ok?

Reply:
Lets raise this with the team leads and see if any teams are actually using it. DAPR is a great tool but it requires ownership and support. DDC would need to be prepared to support the operator and any event grid integration configuration as they take ownership of the AKS clusters. If we're not actively using/supporting DAPR we should probably remove this for now.

Reply:
I'll remove it...

Reply:
Done